### PR TITLE
Code to write UDQ and ACTIONX data to the Eclipse compatible restart file

### PR DESCRIPTION
**Pull request by jalvestad**
This pullrequest contains code to activate the writing of UDQ and ACTIONX related data to the Eclipse compatible restart file. There are also changes to the content of the various vectors that are written out. 

I would like to acknowledge collabration with and contributions to this pullrequest from Joakim Hove and Bård Skaflestad.

### CODE REVIEW

**Comment by joakim-hove** [comment]
@jalvestad - there was a bug in the `Action::Result` copy constructor - I have merged an updated implementation in #1196 - if you rebase your PR on upstream/master things should have improved.

File: opm/output/eclipse/AggregateWellData.hpp [hunk 1]
```diff
@@ -39,6 +40,12 @@ namespace Opm { namespace data {
 }} // Opm::data
 
 namespace Opm { namespace RestartIO { namespace Helpers {
+    
+    struct ActionResStatus {
```
Comment (joakim-hove) [inline opm/output/eclipse/AggregateWellData.hpp]: This is just a small drive-by comment - not important either; but anyway. In situations where you need to create a list of several pieces of information you have (at least) two alternatives:

1. You can create several lists of scalar.
2. You can create one list of a composite object.

In the `ActionResStatus` case you have chosen to create *two* lists of `Action::Result` and `std::string`objects respectively. The disadvantage with this approach is that the modelling/implementation does not focus on the smallest interesting item - namely a pair of `std::string` and `Action::Result`- the item actually does not even exist in the implementation. In my opinion it is generally[1] a good rule to flip things around, and make one of list of objects, in this case its could be as simple as:
```
std::vector<std::pair<std::string, Action::Result>>
```
or maybe:
```
Opm::OrderedMap<std::string, Action::Result>
```
or the more general:
```
struct ActionStatus {
      Opm::Action::Result result;
      std::string name;
}

std::vector<ActionResult> result;
```

[1]: Of course - this is no hard and fast rule, rather an opinion from my side. Obvious exceptions are when one of the vectors - e.g. the list of names in this case - should be used in isolation.

File: opm/output/eclipse/VectorItems/well.hpp [hunk 2]
```diff
@@ -180,6 +180,7 @@ namespace Opm { namespace RestartIO { namespace Helpers { namespace VectorItems
     namespace ZWell {
         enum index : std::vector<const char*>::size_type {
```
Comment (joakim-hove) [inline opm/output/eclipse/VectorItems/well.hpp]: Do you really need the `std::vector<const char*>::size_type` - can't you just go with the default? Anyway good if you use the "new and fancy" `enum class`

File: src/opm/output/eclipse/AggregateUDQData.cpp [hunk 3]
```diff
@@ -456,66 +486,126 @@ captureDeclaredUDQData(const Opm::Schedule&                 sched,
             auto z_udl = this->zUDL_[udq_index];
             zUdl::staticContrib(udq_input, z_udl);
         }
+        cnt_udq += 1;
     }
-
+    if (cnt_udq != nudq) {
+        //std::stringstream str;
+        //str << "Inconsistent total number of udqs: " << cnt_udq << " and sum of well, group and field udqs: " << nudq;
+        //throw std::invalid_argument(str.str());
```
Comment (joakim-hove) [inline src/opm/output/eclipse/AggregateUDQData.cpp]: Some debug code which could be removed?

**Comment by joakim-hove** [comment]
jenkins build this please

File: opm/output/eclipse/VectorItems/well.hpp (hunk 2 above)
Comment (bska) [reply opm/output/eclipse/VectorItems/well.hpp]: I'd advice against making this a strong enums (i.e., `enum class`).  Those cannot be used as indices into a vector without a `static_cast` at the call site.  As for `::size_type`, that's following the pattern elsewhere and is to ensure that no unit conversion happens when we do use these as indices.

**Comment by bska** [comment]
Is this close enough to complete now that we should start the final review and integration build tests?

File: src/opm/output/eclipse/AggregateConnectionData.cpp [hunk 4]
```diff
@@ -191,6 +191,7 @@ namespace {
 
             sConn[Ix::item30] = -1.0e+20f;
             sConn[Ix::item31] = -1.0e+20f;
+            sConn[Ix::item41] = (conn.CF() > 0) ? 1 : 0;
```
Comment (bska) [inline src/opm/output/eclipse/AggregateConnectionData.cpp]: If you rebase this onto current master, which has #1339, you can write this as
```C++
sConn[Ix:item41] = conn.ctfAssignedFromInput() ? 1.0f : 0.0f;
```

**Comment by jalvestad** [comment]
The code is ready for final review and merging into master.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
OK; I have looked at the test failing in opm-common. I am not really familiar with the datastructures used, but I think the problem is that the `dUdw` array has been allocated with wrong size. The crash comes in the assignment:
```C++
template <class DUDWArray>
void staticContrib(const Opm::SummaryState& st,
                           const std::vector<std::string>& wnames,
                           const std::string udq,
                           const std::size_t nwmaxz,
                           DUDWArray&   dUdw)
{
      //initialize array to the default value for the array
      for (std::size_t ind = 0; ind < nwmaxz; ind++) {
           dUdw[ind] = -0.3E+21;
      }
      for (std::size_t ind = 0; ind < wnames.size(); ind++) {
           if (st.has_well_var(wnames[ind], udq)) {
              // The crash comes in the assignment below; look like the dUdw array
              // is created with two small size - something wrong with UDQDims? 
              dUdw[ind] = st.get_well_var(wnames[ind], udq);        
           }
      }
}
```

**Comment by bska** [comment]
> I am not really familiar with the datastructures used, but I think the problem is that the `dUdw` array has been allocated with wrong size.

That's **arguably** a bug in the msim test decks (`tests/msim/*.include`).  Those use
```
WELLDIMS
  2 1 1 2 /
```
(i.e., maximum number of wells in field (item 1) equal to 2), but have at least 4 active wells (5 in `uda.include`).  On the other hand, it looks like we missed this particular item in 7986e99e.  In any case, this problem existed before this PR, but was brought to light here.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Looks like the test failures might be race condition with one of @akva2's PR's - I will rerun.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
Hmmmm - I must admit the number of failing tests was frighteningly large - I would have expected only th UDQ/ACTIONX related tests to fail?

**Comment by bska** [comment]
> I would have expected only the UDQ/ACTIONX related tests to fail?

I suppose this PR started out that way, but during the course of development we discovered other connections that ended up requiring changes to integer data for wells and groups (and headers).  As a result, pretty much every single restart file in the reference data ends up being changed by this PR.

Piecing apart the individual contributions as was attempted in #1358 is probably futile at this point.

**Comment by joakim-hove** [comment]
> As a result, pretty much every single restart file in the reference data ends up being changed by this PR.

OK - that was unfortunate - but then I guess we just have to merge and update data on good faith. By the looks of this PR I am ready to merge it.

**Comment by joakim-hove** [comment]
jenkins build this update_data please

**Comment by joakim-hove** [comment]
jenkins build this opm-tests=217 please

**Comment by bska** [comment]
I would really like to get this into master now.  Any objections?

**Pull request merged**